crates/apollo-compiler/tests/validation/recursion.rs

```diff
@@ -0,0 +1,127 @@
+          }
+        }
+      }
+    "#
+    .to_string();
+
+    for i in 1..size {
+        query.push_str(&format!(
+            "
+          fragment typeFragment{i} on __Type {{
+            ofType {{
+              ...typeFragment{}
+            }}
+          }}",
+            i + 1
+        ));
+    }
+    query.push_str(&format!(
+        "
+          fragment typeFragment{size} on __Type {{
+            ofType {{
+              name
+            }}
+          }}"
+    ));
+
+    query
+}
+
+fn build_directive_chain(size: usize) -> String {
+    let mut schema = r#"
+      type Query {
+        field: Int! @directive(arg: true)
+      }
+
+      directive @directive(arg: Boolean @argDir1) on FIELD_DEFINITION
+    "#
+    .to_string();
+
+    for i in 1..size {
+        schema.push_str(&format!(
+            "
+            directive @argDir{i}(arg: Boolean @argDir{}) on ARGUMENT_DEFINITION
+            ",
+            i + 1
+        ));
+    }
+    schema.push_str(&format!(
+        "
+        directive @argDir{size}(arg: Boolean) on ARGUMENT_DEFINITION
+          "
+    ));
+
+    schema
+}
+
+#[test]
+fn long_fragment_chains_do_not_overflow_stack() {
+    // Build a query that applies 1K fragments
+    // Validating it would take a lot of recursion and blow the stack
+    let query = build_fragment_chain(1_000);
+
+    let errors = apollo_compiler::parse_mixed_validate(
+        format!(
+            "type Query {{ a: Int }}
+            {query}"
+        ),
+        "overflow.graphql",
+    )
+    .expect_err("must have recursion errors");
+
+    let expected = expect_test::expect![[r#"
+        Error: too much recursion
+        Error: `typeFragment1` fragment cannot reference itself
+            ╭─[overflow.graphql:11:11]
+            │
+         11 │           fragment typeFragment1 on __Type {
+            │           ───────────┬──────────  
+            │                      ╰──────────── recursive fragment definition
+        ────╯
+    "#]];
+    expected.assert_eq(&errors.to_string_no_color());
+}
+
+#[test]
+fn not_long_enough_fragment_chain_applies_correctly() {
+    // Stay just under the recursion limit
+    let query = build_fragment_chain(200);
+
+    let _ = apollo_compiler::parse_mixed_validate(
+        format!(
+            "type Query {{ a: Int }}
+            {query}"
+        ),
+        "no_overflow.graphql",
+    )
+    .expect("must not have recursion errors");
+}
+
+#[test]
+fn long_directive_chains_do_not_overflow_stack() {
+    // Build a query that applies 1K directives
+    // Validating it would take a lot of recursion and a lot of time
+    let schema = build_directive_chain(1_000);
+
+    let partial = apollo_compiler::Schema::parse_and_validate(schema, "directives.graphql")
+        .expect_err("must have recursion errors");
+
+    // The final 199 directives do not cause recursion errors because the chain is less than 200
+    // directives deep.
+    assert_eq!(partial.errors.len(), 801);
```

Comment:
Oh, do we restart following the whole chain for every step in it? Quadratic time is why it took 60 seconds… Maybe this limit should be much lower still

Reply:
We dooo. Fragment definitions are validated when they are used, but for directives and input objects each definition is validated.

Long fragment chains do happen in the wild, but long input object chains and directive chains might not, so we could have a low limit for those. The risk is lower for directives and input objects because they are in the schema and not likely to be provided over the internet by a malicious user.